File: src/ts/terrain/landscape/mountainLayer.ts
import { LVector3 } from "../../utils/algebra";
import { minimumValue, pow, sAbs } from "../../utils/gradientMath";
import { simplex411 } from "../../utils/simplex";
import { simpleElevationFunction } from "./elevationFunction";

export function mountainLayer(frequency: number, nbOctaves: number, decay: number, lacunarity: number, power: number, minValue: number): simpleElevationFunction {
    return function (coords: LVector3, seed: number, gradient: LVector3) {
        let noiseValue = 0.0;
        let totalAmplitude = 0.0;
        const localGradient = LVector3.Zero();
        const samplePoint = coords.scale(frequency);
        for (let i = 0; i < nbOctaves; i++) {
            let localElevation = simplex411(samplePoint, seed, localGradient);

            // TODO: ne pas hardcoder
            const sharpness = 8.0;
            localElevation = 1.0 - sAbs(localElevation, sharpness, localGradient);

            localGradient.divideInPlace(-1);

            localElevation /= decay ** i * (1.0 + gradient.getSquaredMagnitude());
            localGradient.divideInPlace(decay ** i * (1.0 + gradient.getSquaredMagnitude()));

            noiseValue += localElevation;
            gradient.addInPlace(localGradient);

            totalAmplitude += 1.0 / decay ** i;

            samplePoint.scaleInPlace(lacunarity);
        }
        noiseValue /= totalAmplitude;
        gradient.divideInPlace(totalAmplitude);

        if (minValue > 0) noiseValue = minimumValue(noiseValue, minValue, gradient);

        return pow(noiseValue, power, gradient);
    };
}

File: packages/terrain-generation/src/landscape/mountain_layer.rs
use crate::utils::math::{s_abs, s_max_gradient};
use crate::utils::simplex::simplex411;
use crate::utils::vector3::Vector3;

pub fn mountain_layer(
    frequency: f32,
    nb_octaves: i32,
    decay: f32,
    lacunarity: f32,
    threshold: f32,
) -> impl Fn(&Vector3, f32, &mut Vector3) -> f32 {
    move |coords: &Vector3, seed: f32, gradient: &mut Vector3| {
        let mut noise_value = 0.0;
        let mut total_amplitude = 0.0;
        let mut local_gradient = Vector3::zero();
        for i in 0..nb_octaves {
            let local_frequency = frequency * f32::powi(lacunarity, i);
            let mut local_elevation =
                simplex411(&(coords * local_frequency), seed, &mut local_gradient);
            local_gradient *= local_frequency;

            // TODO: ne pas hardcoder
            let sharpness = 4.0;
            local_elevation = 1.0 - s_abs(local_elevation, sharpness, &mut local_gradient);

            local_gradient *= -1.0;

            local_elevation /= f32::powi(decay, i);
            local_gradient /= f32::powi(decay, i);

            noise_value += local_elevation;
            *gradient += &local_gradient;

            total_amplitude += 1.0 / f32::powi(decay, i);
        }

        noise_value /= total_amplitude;
        *gradient /= total_amplitude;

        //pow(noise_value, power, gradient)

        noise_value = s_max_gradient(
            noise_value,
            threshold,
            10.0,
            gradient,
            &mut Vector3::new(0.0, 0.0, 0.0),
        );

        noise_value -= threshold;

        noise_value /= 1.0 - threshold;
        *gradient /= 1.0 - threshold;

        noise_value
    }
}
